canard.c

```diff
@@ -132,6 +132,7 @@ uint8_t canardGetLocalNodeID(const CanardInstance* ins)
 }
 
 int16_t canardBroadcast(CanardInstance* ins,
+                        CanardTransportProtocol protocol,
```

Comment:
I propose we set the protocol once during initialization (new argument for `canardInit()`). There is no point in selecting the protocol more than once at runtime, is there? Having it specified once simplifies both the API and internal calls, e.g. `enqueueTxFrames()`.

Reply:
This is required for CAN2.0B and CAN-FD to co-exists with the same interface. You need to set the one you want libcanard to generate frames for. 

I cannot see any other way to get broadcast with both frames as an option to work

Reply:
> This is required for CAN2.0B and CAN-FD to co-exists with the same interface. You need to set the one you want libcanard to generate frames for.

True, but why would one want to set it more than once (i.e., after initialization)? The use case I imagine is as follows: an integrator/engineer takes a third-party UAVCAN node and integrates it into their vehicle. Say, the node utilizes libcanard and supports CAN FD (and therefore CAN 2.0, too). The factory-default configuration tells the node to use CAN FD (because it's superior), but the vehicle it is being integrated into uses CAN 2.0, so the engineer connects to the node and changes its configuration from CAN FD to CAN 2.0. That's it.

This is the main use case I had in mind when we were talking about runtime protocol selection. Did you have a different use case in mind?

Reply:
Yes. Let's say you have a CAN-FD compatible UAVCAN node in a network which are communicating with other CAN-FD compatible UAVCAN nodes. They will, therefore, encode all UAVCAN messages into CAN-FD frames up to 64 Bytes. However, a node with a CAN2.0 controller may still coexist on the same network _as long as the transceiver is CAN-FD tolerant_. When receiving a CAN-FD frame the transceiver will simply "ignore" it and not pass it forward to the controller. When a UAVCAN node is configured to operate in CAN-FD compatible mode it should still be possible to transmit UAVCAN messages to other non-CAN-FD compatible nodes which may exist on the same network. 

It is, therefore, required to select between using CAN-FD or CAN2.0 in runtime.

Reply:
The dynamic selection will be useful only if all of the following criteria are met:

- there are CAN 2.0 and CAN FD nodes mixed inside the same network
- all CAN 2.0 nodes use CAN FD capable drivers
- CAN FD capable nodes know which nodes on the bus are restricted to CAN 2.0

The cost of the dynamic selection feature is a marginal increase in the complexity of the library API a minor increase in the ROM footprint.

My assessment is that the described use case would be exceedingly rare to encounter in the field, in which case the cost of supporting it may make it unworthwhile. Would you not agree?

Reply:
My original idea was to only select this at initialization. With this approach selecting block size at initialization is the clearly superior alternative. There are practically no drawbacks. 

It seemed to me that your problem with selecting block size at initialization was that CAN-FD and CAN2.0 could not co-exist on the same interface at the same time. Did i misunderstand you?

Reply:
I went back and re-read our original discussion and I see that I could be misunderstood. It was never my intention to advocate for protocol selection on a per-transfer basis; rather, whenever I mentioned "runtime selection" I meant selection during initialization, as opposed to compile-time-only. It is true that one can imagine various use cases where such flexibility may be desirable, but they seem of marginal importance.

> With this approach selecting block size at initialization is the clearly superior alternative.

Maybe not? Runtime block selection forces us to do more computation at run time than necessary (I am referring to frequent references to `ins->block_size`) and brings issues with flex array members inside `CanardTxQueueItem`. Increased memory footprint doesn't bother me as much.

Also, larger blocks selected statically will allow us to remove these bit fields in favor of regular byte-aligned fields, possibly improving CPU overhead and eliminating the maximum transfer size limit `CANARD_MAX_TRANSFER_PAYLOAD_LEN` (which is currently set to 1024, which may be too restrictive for CAN FD networks):

https://github.com/UAVCAN/libcanard/blob/4df2678e5b231f047f7a4a069fe0d5988aa7d1bc/canard.h#L241-L245

Here is my extended feature table that should emphasize what I am after:

Capability  | Compiled & inited in FD mode | Compiled in FD mode, inited in CAN2 mode | Compiled & inited in CAN2 mode
------------|------------------------------|------------------------------------------|--------------------------------
Transmission| CAN FD only                  | CAN 2.0 only                             | CAN 2.0 only
Reception   | Any                          | Any                                      | CAN 2.0 only
Pool block size| 88 bytes                  | 88 bytes                                 | 32 bytes
Max payload size| unlimited (4 GB)         | unlimited                                | 1 KiB

Reply:
The flex array members (the one that broke the c specc) is easily fixable. Flex array members themselves should be fine to use in this type of application.

The extra operations you're talking about is basically one subtract operation. I don't think there will be any significant difference, if you're very concerned about this we must benchmark it.

Setting block size at init allows us to only produce and test one binary. And doesn't mess up memory usage when compiling with fd support. 

The only other decent alternative is compiling for fd only or 2.0 only. Then there is no dynamic block size, but we must test two binaries. 

Im against implementing compatibility mode if we're not aiming for simultanous use of different frames. Which i didn't think we needed in the first place.

Reply:
The flex array members (the one that broke the c specc) is easily fixable. Flex array members themselves should be fine to use in this type of application.

The extra operations you're talking about is basically one subtract operation. I don't think there will be any significant difference, if you're very concerned about this we must benchmark it.

Setting block size at init allows us to only produce and test one binary. And doesn't mess up memory usage when compiling with fd support. 

The only other decent alternative is compiling for fd only or 2.0 only. Then there is no dynamic block size, but we must test two binaries. 

Im against implementing compatibility mode if we're not aiming for simultanous use of different frames. Which i didn't think we needed in the first place.

Reply:
Performance is on my list of concerns but only barely, just like memory usage.

Would I be correct at describing your proposed approach with the following table?

Capability      | Inited in CAN FD mode  | Inited in CAN2 mode
----------------|--------------------|----------------------
Transmission    | CAN FD only        | CAN 2.0 only
Reception       | Any                | Any
Pool block size | 88 bytes           | 32 bytes
Max payload size| 1 KiB              | 1 KiB

Observe that the ability to emit only CAN 2.0 while being able to receive both CAN 2.0 and CAN FD is also paramount for bootloaders, which must be able to operate while not knowing whether the network they are attached to is CAN FD or CAN 2.0.

Reply:
> Would I be correct at describing your proposed approach with the following table?

Yes, this is what I had in mind.

> Observe that the ability to emit only CAN 2.0 while being able to receive both CAN 2.0 and CAN FD is also paramount for bootloaders, which must be able to operate while not knowing whether the network they are attached to is CAN FD or CAN 2.0.

What if someone connects them to I2C? I don't really get why bootloaders are special in this regard.

I think configuring this dynamically would be the sensible approach. Selecting block size dynamically allows configuring at run-time between CAN2.0B and CAN-FD without wasting resources.

---

I still feel dynamically chosen block size at initialization is the technically superior alternative. But the "chosen at compile time" alternative is still OK with me.

If we go this route there is a lot of inconveniences we don't really need to have, some of them are:
 - Cannot configure a node between 2.0 and FD at run time (not even bootloaders), without compiling both versions of the library.
 - Cannot connect to multiple buses with different protocols without compiling both versions of the library.
 - A vendor cannot support CAN2.0 and CAN-FD without compiling both versions of the library.

Remember that my original PR was a WIP, and a lot of the weirdness seen in it can be removed.

Reply:
> What if someone connects them to I2C? I don't really get why bootloaders are special in this regard.

Bootloaders are special because you can't configure them. A bootloader must be able to operate on a bus while having zero prior knowledge of its properties (other than it's UAVCAN-compliant) - whether it's CAN 2.0 or CAN FD, its bit rate, the node ID of the local device, etc. A bootloader that is dependent on correct configuration complicates the update process and makes the system far less brick-proof than it could be. We researched this question with the PX4 team a few years back while building the [PX4 Brickproof Bootloader](https://dev.px4.io/en/uavcan/bootloader_installation.html).

Regardless, though, if we are able to receive both CAN FD and CAN 2.0 while operating in the CAN 2.0 mode (which seems to be the case in all of the discussed alternatives so far), we're okay and there is nothing more to discuss.

> If we go this route there is a lot of inconveniences we don't really need to have, some of them are:

Surely we can switch between CAN FD and CAN 2.0 at run time (e.g., during initialization) if the block size is large enough. Selecting CAN 2.0 in that case comes with suboptimal memory usage, as we discussed earlier. If we are okay with wasting memory, none of the listed inconveniences seem to exist.

I am not totally against the runtime-selectable block size. If you want it done this way, it certainly can be; we can go back to the original PR.